### PR TITLE
SQLAlchemy Persistence on Python Applications

### PR DESCRIPTION
**Pull request by brunokrebs**

### CODE REVIEW

File: _posts/2017-10-20-sqlalchemy-orm-tutorial-for-python-developers.markdown [hunk 1]
```diff
@@ -0,0 +1,688 @@
...
+
+### SQLAlchemy ORM Cascade
+
+Whenever rows in a particular table are updated or deleted, rows in other tables might need to suffer changes as well. These changes can be simple updates, which are called cascade updates, or full deletes, known as cascade deletes. For example, let's say that we have a table called `shopping_carts`, a table called `products`, and a third one called `shopping_carts_products` that connects the first two tables. If, for some reason, we need to delete rows from `shopping_carts` we will need to delete the related rows from `shopping_carts_products` as well. Otherwise we will end up with a lot of garbage and unfulfilled references in our database.
+
+To make this kind of operation easily to maintain, SQLAlchemy ORM enables developers to map cascade behavior when using `relationship()` constructs. Like that, when operations are performed on _parent_ objects, _child_ objects get updated/deleted as well. The following list provides a brief explanation of the most used cascade strategies on SQLAlchemy ORM:
```
Comment (unicodeveloper) [inline _posts/2017-10-20-sqlalchemy-orm-tutorial-for-python-developers.markdown]: easily => easy

File: _posts/2017-10-20-sqlalchemy-orm-tutorial-for-python-developers.markdown [hunk 2]
```diff
@@ -0,0 +1,688 @@
...
+
+# create a Session
+session = Session()
+```
+
+As we can see from the code snippet above, we only need one step to get sessions. We need to create a session factory that is bound to the SQLAlchemy engine. After that, we can just issue calls to this session factory to get our sessions.
+
+## SQLAlchemy in Practice
+
+Now that we got a better understanding of the most important pieces of SQLAlchemy, it's time to start practicing it. In the following sections, we will create a small project based on [`pipenv`](https://github.com/kennethreitz/pipenv)—a Python dependency manager—and add some classes to it. Then we will map these classes to tables persisted on a PostgreSQL database and learn how to query data.
```
Comment (unicodeveloper) [inline _posts/2017-10-20-sqlalchemy-orm-tutorial-for-python-developers.markdown]: persisted on a PostgreSQL database =>
persisted to a PostgreSQL database

File: _posts/2017-10-20-sqlalchemy-orm-tutorial-for-python-developers.markdown [hunk 3]
```diff
@@ -0,0 +1,688 @@
...
+# stop instance
+docker stop sqlalchemy-orm-psql
+
+# destroy instance
+docker rm sqlalchemy-orm-psql
+```
+
+The first command, the one that creates the PostgreSQL instance, contains a few parameters that are worth inspecting:
+
+- `--name`: Defines the name of the Docker instance
```
Comment (unicodeveloper) [inline _posts/2017-10-20-sqlalchemy-orm-tutorial-for-python-developers.markdown]: Please add a full stop to the sentence.

File: _posts/2017-10-20-sqlalchemy-orm-tutorial-for-python-developers.markdown [hunk 4]
```diff
@@ -0,0 +1,688 @@
...
+- `-e POSTGRES_PASSWORD`: Defines the password to connect to PostgreSQL.
+- `-e POSTGRES_USER`: Defines the user to connect to PostgreSQL.
+- `-e POSTGRES_DB`: Defines the main (and only) database available in the PostgreSQL instance.
+- `-p 5432:5432`: Defines that the local `5432` port will tunnel connections to the same port in the Docker instance.
+- `-d postgres`: Defines that this Docker instance will be created based on the [official PostgreSQL repository](https://hub.docker.com/_/postgres/).
+
+### Installing SQLAlchemy Dependencies
+
+In this tutorial, we will need to install only two packages: `sqlalchemy` and `psycopg2`. The first dependency refers to SQLAlchemy itself and the second one, `psycopg2`, is the PostgreSQL driver that SQLAlchemy will use to communicate with the database. To install these dependencies, we will use `pipenv` as shown:
```
Comment (unicodeveloper) [inline _posts/2017-10-20-sqlalchemy-orm-tutorial-for-python-developers.markdown]: Link to these packages?

File: _posts/2017-10-20-sqlalchemy-orm-tutorial-for-python-developers.markdown [hunk 5]
```diff
@@ -0,0 +1,688 @@
...
+Base = declarative_base()
+```
+
+This code creates:
+
+- a SQLAlchemy Engine that will interact with our _dockerized_ PostgreSQL database,
+- a SQLAlchemy ORM session factory bound to this engine,
+- and a base class for our classes definitions.
+
+Now let's create and map the `Movie` class. To do this, let's create a new file called `movie.py` and add to it the following code:
```
Comment (unicodeveloper) [inline _posts/2017-10-20-sqlalchemy-orm-tutorial-for-python-developers.markdown]: and add to it the following code: =>

and add the following code to it.

File: _posts/2017-10-20-sqlalchemy-orm-tutorial-for-python-developers.markdown [hunk 6]
```diff
@@ -0,0 +1,688 @@
...
+session.add(matt_stuntman)
+session.add(dwayne_stuntman)
+session.add(mark_stuntman)
+
+# 10 - commit and close session
+session.commit()
+session.close()
+```
+
+This code is split into 10 sections. Let's inspecting them:
```
Comment (unicodeveloper) [inline _posts/2017-10-20-sqlalchemy-orm-tutorial-for-python-developers.markdown]: inspecting => inspect.

File: _posts/2017-10-20-sqlalchemy-orm-tutorial-for-python-developers.markdown [hunk 7]
```diff
@@ -0,0 +1,688 @@
...
+movies = session.query(Movie).all()
+
+# 4 - print movies' details
+print('\n### All movies:')
+for movie in movies:
+    print(f'{movie.title} was released on {movie.release_date}')
+print('')
+```
+
+The code snippet above—that can be run with `python queries.py`,—shows how easy it's to use SQLAlchemy ORM to query data. To retrieve all movies from the database, we just needed to fetch a session from the session factory, use it to get a query associated with `Movie`, and then call the `all()` function on this query object. The Query API provides dozens of useful functions like `all()`. In the following list, we can see a brief explanation about the most important ones:
```
Comment (unicodeveloper) [inline _posts/2017-10-20-sqlalchemy-orm-tutorial-for-python-developers.markdown]: it's => it is

File: _posts/2017-10-20-sqlalchemy-orm-tutorial-for-python-developers.markdown [hunk 8]
```diff
@@ -0,0 +1,688 @@
...
+    .filter(ContactDetails.address.ilike('%glendale%')) \
+    .all()
+
+print('### Dwayne Johnson movies:')
+for actor in glendale_stars:
+    print(f'{actor.name} has a house in Glendale')
+print('')
+```
+
+The fifth section of the updated script uses the `filter()` function to fetch only movies that were released after January the first, 2015. The sixth section shows how to use `join()` to fetch instances of `Movie` that the `Actor` Dwayne Johnson participated in. The seventh and last section, shows the usage of `join()` and `ilike()` functions to retrieve the actors that have house in Glendale.
```
Comment (unicodeveloper) [inline _posts/2017-10-20-sqlalchemy-orm-tutorial-for-python-developers.markdown]: house or houses?

File: _posts/2017-10-20-sqlalchemy-orm-tutorial-for-python-developers.markdown [hunk 9]
```diff
@@ -0,0 +1,688 @@
...
+
+Before diving in the ORM features provided by SQLAlchemy, we need to learn how the core works. The following sections will introduce important concepts that every Python developer needs to understand before dealing with SQLAlchemy applications.
+
+### Python DBAPI
+
+The [Python DBAPI (an acronym for DataBase API)](https://www.python.org/dev/peps/pep-0249/) was created to specify how Python modules that integrate with databases should expose their interfaces. Although we won't interact with this API directly—we will use SQLAlchemy as a facade to it—it's good to know that it defines how common functions like `connect`, `close`, `commit`, and `rollback` must behave. Consequently, whenever we use a Python module that adheres to the specification, we can rest assured that we will find these functions and that they will behave as expected.
```
[the remark this answers is not in the document]
Comment (brunokrebs) [reply _posts/2017-10-20-sqlalchemy-orm-tutorial-for-python-developers.markdown]: They are called em dashes actually. http://www.thepunctuationguide.com/em-dash.html

File: _posts/2017-10-20-sqlalchemy-orm-tutorial-for-python-developers.markdown (hunk 4 above)
Comment (brunokrebs) [reply _posts/2017-10-20-sqlalchemy-orm-tutorial-for-python-developers.markdown]: Added on previous sections.

**Pull request merged**